src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultImportDocumentManager.cs

```diff
@@ -0,0 +1,235 @@
+            {
+                var importFilePath = import.PhysicalPath;
+                Debug.Assert(importFilePath != null);
+
+                if (!_importTrackerCache.TryGetValue(importFilePath, out var importTracker))
+                {
+                    // First time seeing this import. Start tracking it.
+                    importTracker = new ImportTracker(importFilePath);
+                    _importTrackerCache[importFilePath] = importTracker;
+
+                    StartListeningForChanges(importTracker);
+                }
+
+                importTracker.AssociatedDocuments.Add(tracker.FilePath);
+            }
+        }
+
+        public override void OnUnsubscribed(VisualStudioDocumentTracker tracker)
+        {
+            if (tracker == null)
+            {
+                throw new ArgumentNullException(nameof(tracker));
+            }
+
+            _foregroundDispatcher.AssertForegroundThread();
+
+            var imports = GetImportItems(tracker);
+            foreach (var import in imports)
+            {
+                var importFilePath = import.PhysicalPath;
+                Debug.Assert(importFilePath != null);
+
+                if (_importTrackerCache.TryGetValue(importFilePath, out var importTracker))
+                {
+                    importTracker.AssociatedDocuments.Remove(tracker.FilePath);
+
+                    if (importTracker.AssociatedDocuments.Count == 0)
+                    {
+                        // There are no open documents that care about this import. We no longer need to track it.
+                        StopListeningForChanges(importTracker);
+                        _importTrackerCache.Remove(importFilePath);
+                    }
+                }
+            }
+        }
+
+        private IEnumerable<RazorProjectItem> GetImportItems(VisualStudioDocumentTracker tracker)
+        {
+            var projectDirectory = Path.GetDirectoryName(tracker.ProjectPath);
+            var templateEngine = _templateEngineFactoryService.Create(projectDirectory, _ => { });
+            var imports = templateEngine.GetImportItems(tracker.FilePath);
+
+            return imports;
+        }
+
+        private void FireImportChanged(string importPath, ImportChangeKind kind)
+        {
+            var handler = Changed;
+            if (handler != null && _importTrackerCache.TryGetValue(importPath, out var importTracker))
+            {
+                var args = new ImportsChangedEventArgs(importPath, kind, importTracker.AssociatedDocuments);
+                handler(this, args);
+            }
+        }
+
+        private void OnFilesChanged(uint fileCount, string[] filePaths, uint[] fileChangeFlags)
+        {
+            for (var i = 0; i < fileCount; i++)
+            {
+                var kind = ImportChangeKind.Changed;
+                var flag = (_VSFILECHANGEFLAGS)fileChangeFlags[i];
+
+                if (flag.HasFlag(_VSFILECHANGEFLAGS.VSFILECHG_Del))
```

Comment:
don't use `HasFlag`, it does a ton of extra work under the covers. Use:

`if ((flag & _VSFILECHANGEFLAGS.VSFILECHG_Del) == _VSFILECHANGEFLAGS.VSFILECHG_Del)`

Same below